### ISSUE TITLE
Prisma migrate feature request: Control binary download and network access by CLI

### ISSUE THREAD

**Issue by LumaKernel** [prisma/prisma#5105]
## Problem

1. Prisma CLI needs write access for non-existent binaries. I'm using `PRISMA_CLI_BINARY_TARGETS="rhel-openssl-1.0.x" npx prisma generate` beforehand, but it may not be documented yet anywhere. I'm not desiring installing native binary for decreasing the size of shipping.
2. Migrate deployments always need network (Internet) access for checking digest sum even if the binary exists.

```
Error: request to https://binaries.prisma.sh/all_commits/5d491261d382a2a5ffdc71de17072b0e409f1cc1/linux-musl/query-engine.gz.sha256 failed, reason: getaddrinfo EAI_AGAIN binaries.prisma.sh
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

### No write accesss

`chown -R -w node_modules`

### No network

```
docker run --rm -it --network none -v $PWD:/tmp/work node:14-alpine /bin/ash
cd /tmp/work
npx prisma deploy --preview-feature
```

## Suggested solution

1. Add them to document, or add subcommand like `prisma download rhel-openssl-1.0.x`
2. Create options like `--skip-check-sum`, `--no-network`

## Additional context

It's also nice to set up the route to the Internet, but I'm downloading the binary by `prisma generate` and carrying it out to the environment right before running migration, so I wish this could be done. ( And it costs setting up just for this purpose... )

Related: https://github.com/prisma/prisma/issues/1383

I encountered the problem while using the serverless model and migration in private network.

Thank you.

**Comment by albertoperdomo** [comment]
It's a general CLI behavior about how we handle binaries, not specific to Migrate.
But this is specific to the deployment process when you are trying to migrate the database.

**Comment by matthewmueller** [comment]
Does specifying the `binaryTargets` in the generator block not work for you? I think it adds around 10mb overhead.

https://www.prisma.io/docs/concepts/components/prisma-schema/generators

This is meant to address the serverless use-case of building for multiple platforms locally. Something like:

```
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "rhel-openssl-1.0.x"]
  previewFeatures = ["sample-preview-feature"]
}
```

**Comment by LumaKernel** [comment]
My bad, little confused. `PRISMA_CLI_BINARY_TARGETS=...` is not needed. I'm using this for saving binary size, but I finally chose to remove after downloading combining `binaryTargets` in `schema.prisma`.

- All prisma commands →  `node_modules/@prisma/engine/{query-engine,prisma-fmt,introspection,migration-engine}-${target}`
  - Needs all of them to be present. ( I also think it's nice make unnecessary binaries removable for production. They except `migration-engine` occupy near 80MB. )
  - **Needs Internet access even if present**.
- `prisma generate` → `node_modules/.prisma/client/query-engine-${target}`
  - I'm running this beforehand.

Please note that Lambda Function unzipped size should be less than around 250MB.

Now, 2 main requests.

- Feature to skip trying to check sum while running `prisma ${command}`.
- In addition, Feature to skip trying to download absent binaries while running `prisma ${command}`.

Optionally, following might be useful.

- Feature to specify targets.
- Feature to specify binary names.

Whether to split migration and others or not, those features might be useful.

**Comment by rostislav-simonik** [comment]
The same, if we specify `binaryTargets = ["native", "rhel-openssl-1.0.x"]`  in schema.prisma, then the `prisma generate` used in serverless config downloads and bundles also native version.

I solved it by adding another command in serverless config after `prisma generate` to remove undesired binaries. 
![image](https://user-images.githubusercontent.com/25525736/109397143-836b3180-7935-11eb-82ba-5bc94767035f.png)

But would be great if `prisma generate` cli option could override `binaryTarges` value present  in `schema.prisma`

**Comment by Jolg42** [comment]
@rostislav-simonik we actually have a currently undocumented way to select which binaries needs to be installed for the CLI
Check it here https://github.com/prisma/docs/issues/1152

Happy to get your feedback if that helps or not 😃

**Comment by James-Quigley** [comment]
I'd like to second a request to enable something like `--skip-check-sum`, `--no-network`. I have a usecase where the post install doesn't have access to the outside internet and therefore fails because it can't download the checksum.

**Comment by clayrisser** [comment]
@James-Quigley yes I also have a situation where I need to run this without internet. `--no-network` would be great

**Comment by thw0rted** [comment]
@Jolg42 any idea why your cache is in your home directory, rather than under the consuming project's `node_modules`?  It looks like the cache path is determined by https://www.npmjs.com/package/find-cache-dir:

> The algorithm checks for the CACHE_DIR environmental variable and uses it if it is not set to true, false, 1 or 0. If one is not found, it tries to find a package.json file, searching every parent directory of the cwd specified (or implied from other options). It returns a string containing the absolute path to the cache directory, or undefined if package.json was never found or if the node_modules directory is unwritable.

I can't figure out how that would wind up with your home directory unless you set `CACHE_DIR` explicitly.  The good news is, that environment variable should make it possible to specify a place to put the cache, then ADD it in your Dockerfile.  I'm trying to solve a similar problem now -- avoiding network requests after building an application in CI -- and I'm going to see if including the cache will help.

**Comment by imorph** [comment]
Did not found this anywhere so will post here.

I want to be able to run migrations inside disconnected environment so download binaries at the time of `migrate deploy` is not an option. 

What worked for me:
When building docker image for production not only install prisma cli, but also run `npx prisma -v` this will download all dependencies for your `PRISMA_CLI_BINARY_TARGETS` at docker-build time.

```Dockerfile
ENV PRISMA_CLI_BINARY_TARGETS=linux-musl-openssl-3.0.x
RUN npm install -g prisma && npx prisma -v
```

also make note that `PRISMA_CLI_BINARY_TARGETS` should point only to one your target (that one running in prod) `PRISMA_CLI_BINARY_TARGETS=native,linux-musl-openssl-3.0.x` for some reasons multiple defs cause problems

hope this helps

**Comment by thw0rted** [comment]
@imorph did you not also run into https://github.com/prisma/prisma/issues/12230 ?  That's a handy trick, but I do wonder if `prisma -v` also caches all the required checksums for you.  (If it does, it would really help with #12230 ...)

**Comment by imorph** [comment]
>  I do wonder if prisma -v also caches all the required checksums for you

@thw0rted I can't confirm that, but my problem was manifesting itself exactly with `*.sha` files (timeout when downloading) and it is gone after i added `prisma -v` to my Dockerfile, so I believe #12230 also will be fixed with this trick
